pkg/aggregator/internal/limiter/limiter.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package limiter
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagset"
+)
+
+type entry struct {
+	current  int // number of contexts currently in aggregator
+	rejected int // number of rejected samples
+	tags     []string
+}
+
+// Limiter tracks number of contexts based on origin detection metrics
+// and rejects samples if the number goes over the limit.
+//
+// Not thread safe.
+type Limiter struct {
+	key   string
+	tags  []string
+	limit int
+	usage map[string]*entry
+}
+
+// New returns a new instance of limiter.
+//
+// If limit is zero or less the limiter is disabled.
+func New(limit int, key string, tags []string) *Limiter {
+	if limit <= 0 {
+		return nil
+	}
+
+	if !strings.HasSuffix(key, ":") {
+		key += ":"
```

Comment:
This ensures that we don't accidentally match tag prefixes when looking for the unique tag id. ie, if we specify that the sender key is `pod_name` and there are two tags present on some future context, `pod_name_alias` and `pod_name`, we only match on `pod_name`